playground/backend/cmd/server/controller.go

```diff
@@ -237,58 +237,138 @@ func setupValidators(sdk pb.Sdk, filepath string) *[]validators.Validator {
+			return
+		}
 		processSuccess(ctx, nil, pipelineId, cacheService, pb.Status_STATUS_COMPILING)
 	}
 
 	// compile
 	logger.Infof("%s: Compile() ...\n", pipelineId)
-	compileCmd := exec.Compile()
-	if data, err := compileCmd.CombinedOutput(); err != nil {
-		processError(ctx, err, data, pipelineId, cacheService, pb.Status_STATUS_COMPILE_ERROR)
+	compileCmd := exec.Compile(ctxWithTimeout)
+	go func(doneCh chan bool, errCh chan error, dataCh chan interface{}) {
+		data, err := compileCmd.CombinedOutput()
+		dataCh <- data
+		if err != nil {
+			errCh <- err
+			doneCh <- false
+		} else {
+			doneCh <- true
+		}
+	}(doneChannel, errorChannel, dataChannel)
+
+	select {
+	case <-ctxWithTimeout.Done():
+		finishByContext(ctxWithTimeout, pipelineId, cacheService)
 		return
-	} else {
-		processSuccess(ctx, data, pipelineId, cacheService, pb.Status_STATUS_EXECUTING)
+	case ok := <-doneChannel:
+		data := <-dataChannel
+		if !ok {
+			err := <-errorChannel
+			processError(ctxWithTimeout, err, data.([]byte), pipelineId, cacheService, pb.Status_STATUS_COMPILE_ERROR)
+			return
+		}
+		processSuccess(ctxWithTimeout, data.([]byte), pipelineId, cacheService, pb.Status_STATUS_EXECUTING)
 	}
 
 	runBuilder, err := setupRunBuilder(pipelineId, lc, sdk, env, compileBuilder)
 	if err != nil {
 		logger.Errorf("%s: error during setup runBuilder: %s\n", pipelineId, err.Error())
-		setToCache(ctx, cacheService, pipelineId, cache.Status, pb.Status_STATUS_ERROR)
+		setToCache(ctxWithTimeout, cacheService, pipelineId, cache.Status, pb.Status_STATUS_ERROR)
 		return
 	}
 
 	// build executor for run step
 	exec = runBuilder.Build()
 
+	// run
 	logger.Infof("%s: Run() ...\n", pipelineId)
-	runCmd := exec.Run()
-	if data, err := runCmd.CombinedOutput(); err != nil {
-		processError(ctx, err, data, pipelineId, cacheService, pb.Status_STATUS_ERROR)
+	runCmd := exec.Run(ctxWithTimeout)
+	go func(doneCh chan bool, errCh chan error, dataCh chan interface{}) {
+		data, err := runCmd.CombinedOutput()
+		dataCh <- data
+		if err != nil {
+			errCh <- err
+			doneCh <- false
+		} else {
+			doneCh <- true
+		}
+	}(doneChannel, errorChannel, dataChannel)
+
+	select {
+	case <-ctxWithTimeout.Done():
+		finishByContext(ctxWithTimeout, pipelineId, cacheService)
 		return
-	} else {
-		processSuccess(ctx, data, pipelineId, cacheService, pb.Status_STATUS_FINISHED)
+	case ok := <-doneChannel:
+		data := <-dataChannel
+		if !ok {
+			err := <-errorChannel
+			processError(ctxWithTimeout, err.(error), data.([]byte), pipelineId, cacheService, pb.Status_STATUS_ERROR)
+			return
+		}
+		processSuccess(ctxWithTimeout, data.([]byte), pipelineId, cacheService, pb.Status_STATUS_FINISHED)
 	}
 }
 
+// finishByContext is used in case of runCode method finished by timeout
+func finishByContext(ctx context.Context, pipelineId uuid.UUID, cacheService cache.Cache) {
+	logger.Errorf("%s: processCode finish because of timeout\n", pipelineId)
```

Comment:
can we make sure the log shows where the timeout happened? compile / prepare / verify / run?

Reply:
maybe even print the whole context

Reply:
Now before each step (validate/prepare/…) we have a log string like:
`logger.Infof("%s: Validate() ...\n", pipelineId)` in the `processCode()` method.
At the end of each step we also have a log string like:
`logger.Infof("%s: Validate() finish\n", pipelineId)` in the `processSuccess()` method.
For example:
```
...
{pipelineId}: Validate() ...
{pipelineId}: Validate() finish
{pipelineId}: Prepare() ...
{pipelineId}: processCode finish because of timeout
...
```
So I guess using these logs we could understand which step timeout has occurred.